File: Plugins/Orimath.Basics.View/Controls/IconImageConverter.cs
﻿using System;
using System.Globalization;
using System.IO;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using Orimath.Controls;

namespace Orimath.Basics.View.Controls
{
    [ValueConversion(typeof(Stream), typeof(Image))]
    public class IconImageConverter : IValueConverter
    {
        private readonly BitmapImage _source;

        public IconImageConverter()
        {
            _source = new BitmapImage
            {
                CacheOption = BitmapCacheOption.OnLoad
            };
            _source.BeginInit();
            _source.UriSource = new Uri("pack://application:,,,/Orimath.Basics.View;component/defaultIcon.png");
            _source.EndInit();
        }

        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is Stream stream)
            {
                using (stream)
                {
                    var source = new BitmapImage
                    {
                        CacheOption = BitmapCacheOption.OnLoad
                    };
                    source.BeginInit();
                    source.StreamSource = stream;
                    source.EndInit();

                    return new AutoDisableImage
                    {
                        Source = source,
                        Stretch = Stretch.Uniform,
                        Width = 16.0,
                        Height = 16.0,
                    };
                }
            }
            else
            {
                return new AutoDisableImage
                {
                    Source = _source,
                    Stretch = Stretch.Uniform,
                    Width = 16.0,
                    Height = 16.0,
                };
            }
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
}

File: Plugins/Orimath.Basics.View/Controls/IconImageConverter.fs
namespace Orimath.Basics.View.Controls
open System
open System.IO
open System.Windows.Controls
open System.Windows.Data
open System.Windows.Media
open System.Windows.Media.Imaging
open Orimath.Controls

[<ValueConversion(typeof<Stream>, typeof<Image>)>]
type IconImageConverter() =
    let source = BitmapImage()
    do
        source.CacheOption <- BitmapCacheOption.OnLoad
        source.BeginInit()
        source.UriSource <- Uri("pack://application:,,,/Orimath.Basics.View;component/defaultIcon.png")
        source.EndInit()

    interface IValueConverter with
        member _.Convert(value, _, _, _) =
            let src =
                match value with
                | :? Stream as stream ->
                    use stream = stream
                    let src = BitmapImage()
                    src.CacheOption <- BitmapCacheOption.OnLoad
                    src.BeginInit()
                    src.StreamSource <- stream
                    src.EndInit()
                    src
                | _ -> source
            let result = AutoDisableImage()
            result.Source <- src
            result.Stretch <- Stretch.Uniform
            result.Width <- 16.0
            result.Height <- 16.0
            box result

        member _.ConvertBack(_, _, _, _) = raise (NotImplementedException())
